// @paperclicks/optimizer_database/models/traffic_source_performance/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrafficSourcePerformanceModels = exports.TikTokPerformance = exports.FacebookPerformance = exports.AmazonPerformance = void 0;
const AmazonPerformance_1 = require("./AmazonPerformance");
const FacebookPerformance_1 = require("./FacebookPerformance");
const GoogleAdsPerformance_1 = require("./GoogleAdsPerformance");
const TikTokPerformance_1 = require("./TikTokPerformance");
const YahooDSPPerformance_1 = require("./YahooDSPPerformance");
var AmazonPerformance_2 = require("./AmazonPerformance");
Object.defineProperty(exports, "AmazonPerformance", { enumerable: true, get: function () { return AmazonPerformance_2.AmazonPerformance; } });
var FacebookPerformance_2 = require("./FacebookPerformance");
Object.defineProperty(exports, "FacebookPerformance", { enumerable: true, get: function () { return FacebookPerformance_2.FacebookPerformance; } });
var TikTokPerformance_2 = require("./TikTokPerformance");
Object.defineProperty(exports, "TikTokPerformance", { enumerable: true, get: function () { return TikTokPerformance_2.TikTokPerformance; } });
exports.TrafficSourcePerformanceModels = {
    AmazonPerformance: AmazonPerformance_1.AmazonPerformance,
    FacebookPerformance: FacebookPerformance_1.FacebookPerformance,
    TikTokPerformance: TikTokPerformance_1.TikTokPerformance,
    GoogleAdsPerformance: GoogleAdsPerformance_1.GoogleAdsPerformance,
    YahooDSPPerformance: YahooDSPPerformance_1.YahooDSPPerformance,
};
