// @paperclicks/optimizer_database/core/read/request/GetPerformanceCondition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPerformanceCondition = void 0;
function isSqlFunction(value) {
    return !!value && typeof value === 'object';
}
class GetPerformanceCondition {
    constructor(...[left, operator, right]) {
        this._left = left;
        this._operator = operator;
        this._right = right;
    }
    toString() {
        return `${this._left} ${this._operator} ${JSON.stringify(this.right)}`;
    }
    get left() {
        return this._left;
    }
    get right() {
        return this._right;
    }
    get operator() {
        return this._operator;
    }
    get fields() {
        return [this._left];
    }
    updateQuery(query, having) {
        const parameter = isSqlFunction(this.right)
            ? this.right.value
            : this.right;
        const { query: newQuery, parameter: queryParameter } = query.setParameter(this.left, parameter);
        const right = isSqlFunction(this.right)
            ? `${this.right.sqlFunction}( '${parameter}' )`
            : queryParameter;
        if (having) {
            return newQuery.andHaving(this.left, this.operator, right);
        }
        return newQuery.andWhere(this.left, this.operator, right);
    }
}
exports.GetPerformanceCondition = GetPerformanceCondition;
