// @paperclicks/optimizer_database/models/tracker_performance/GaPerformance.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GaPerformance = exports.GaPerformanceCreate = void 0;
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const PerformanceTable_1 = require("../../tables/PerformanceTable");
const TrackerPerformance_1 = require("../base/TrackerPerformance");
class GaPerformanceCreate extends TrackerPerformance_1.TrackerPerformanceCreate {
    constructor(model) {
        super(model);
    }
}
exports.GaPerformanceCreate = GaPerformanceCreate;
class GaPerformance extends TrackerPerformance_1.TrackerPerformance {
    constructor() {
        super(...arguments);
        this['ga_users'] = 0;
        this['ga_newUsers'] = 0;
        this['ga_sessions'] = 0;
        this['ga_bounces'] = 0;
        this['ga_sessionDuration'] = 0;
        this['ga_adsenseRevenue'] = 0;
        this['ga_adsenseAdUnitsViewed'] = 0;
        this['ga_adsenseAdsViewed'] = 0;
        this['ga_adsenseAdsClicks'] = 0;
        this['ga_adsensePageImpressions'] = 0;
        this['ga_goal1Starts'] = 0;
        this['ga_goal2Starts'] = 0;
        this['ga_goal3Starts'] = 0;
        this['ga_goal4Starts'] = 0;
        this['ga_goal5Starts'] = 0;
        this['ga_goal1Value'] = 0;
        this['ga_goal2Value'] = 0;
        this['ga_goal3Value'] = 0;
        this['ga_goal4Value'] = 0;
        this['ga_goal5Value'] = 0;
        this['ga_goal1Completions'] = 0;
        this['ga_goal2Completions'] = 0;
        this['ga_goal3Completions'] = 0;
        this['ga_goal4Completions'] = 0;
        this['ga_goal5Completions'] = 0;
        this['ga_entrances'] = 0;
        this['ga_pageviews'] = 0;
        this['ga_timeOnPage'] = 0;
        this['ga_pageValue'] = 0;
        this['ga_uniquePageviews'] = 0;
        this['ga_exits'] = 0;
    }
}
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_1'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "view_id", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_2'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_users", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_3'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_newUsers", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_4'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_sessions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_5'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_bounces", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_6'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_sessionDuration", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }) //TODO: Maybe average is needed here
    ,
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_7'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_adsenseRevenue", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_8'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_adsenseAdUnitsViewed", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_9'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_adsenseAdsViewed", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_10'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_adsenseAdsClicks", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_11'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_adsensePageImpressions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_12'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal1Starts", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_13'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal2Starts", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_14'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal3Starts", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_15'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal4Starts", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_16'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal5Starts", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_17'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal1Value", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_18'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal2Value", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_19'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal3Value", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_20'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal4Value", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_21'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal5Value", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_22'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal1Completions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_23'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal2Completions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_24'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal3Completions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_25'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal4Completions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_26'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_goal5Completions", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_27'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_entrances", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_28'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_pageviews", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_29'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_timeOnPage", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_30'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_pageValue", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_31'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_uniquePageviews", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_32'),
    __metadata("design:type", Number)
], GaPerformance.prototype, "ga_exits", void 0);
exports.GaPerformance = GaPerformance;
