// @paperclicks/optimizer_database/models/tracker_performance/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrackerPerformanceModels = exports.VoluumPerformance = exports.PublisherPerformance = exports.GaPerformance = exports.ClickflarePerformance = exports.AYPerformance = void 0;
const AssertiveYieldPerformance_1 = require("./AssertiveYieldPerformance");
const ClickflarePerformance_1 = require("./ClickflarePerformance");
const GaPerformance_1 = require("./GaPerformance");
const PublisherPerformance_1 = require("./PublisherPerformance");
const VoluumPerformance_1 = require("./VoluumPerformance");
var AssertiveYieldPerformance_2 = require("./AssertiveYieldPerformance");
Object.defineProperty(exports, "AYPerformance", { enumerable: true, get: function () { return AssertiveYieldPerformance_2.AYPerformance; } });
var ClickflarePerformance_2 = require("./ClickflarePerformance");
Object.defineProperty(exports, "ClickflarePerformance", { enumerable: true, get: function () { return ClickflarePerformance_2.ClickflarePerformance; } });
var GaPerformance_2 = require("./GaPerformance");
Object.defineProperty(exports, "GaPerformance", { enumerable: true, get: function () { return GaPerformance_2.GaPerformance; } });
var PublisherPerformance_2 = require("./PublisherPerformance");
Object.defineProperty(exports, "PublisherPerformance", { enumerable: true, get: function () { return PublisherPerformance_2.PublisherPerformance; } });
var VoluumPerformance_2 = require("./VoluumPerformance");
Object.defineProperty(exports, "VoluumPerformance", { enumerable: true, get: function () { return VoluumPerformance_2.VoluumPerformance; } });
exports.TrackerPerformanceModels = {
    AYPerformance: AssertiveYieldPerformance_1.AYPerformance,
    ClickflarePerformance: ClickflarePerformance_1.ClickflarePerformance,
    GaPerformance: GaPerformance_1.GaPerformance,
    PublisherPerformance: PublisherPerformance_1.PublisherPerformance,
    VoluumPerformance: VoluumPerformance_1.VoluumPerformance,
};
