// @paperclicks/optimizer_database/models/tracker_performance/VoluumPerformance.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoluumPerformance = exports.VoluumPerformanceCreate = void 0;
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const PerformanceTable_1 = require("../../tables/PerformanceTable");
const TrackerPerformance_1 = require("../base/TrackerPerformance");
class VoluumPerformanceCreate extends TrackerPerformance_1.TrackerPerformanceCreate {
    constructor(model) {
        super(model);
    }
}
exports.VoluumPerformanceCreate = VoluumPerformanceCreate;
class VoluumPerformance extends TrackerPerformance_1.TrackerPerformance {
    constructor() {
        super(...arguments);
        this.customConversions1 = 0;
        this.customConversions2 = 0;
        this.customConversions3 = 0;
        this.customConversions4 = 0;
        this.customConversions5 = 0;
        this.customConversions6 = 0;
        this.customConversions7 = 0;
        this.customConversions8 = 0;
        this.customConversions9 = 0;
        this.customConversions10 = 0;
        this.customConversions11 = 0;
        this.customConversions12 = 0;
        this.customConversions13 = 0;
        this.customConversions14 = 0;
        this.customConversions15 = 0;
        this.customConversions16 = 0;
        this.customConversions17 = 0;
        this.customConversions18 = 0;
        this.customConversions19 = 0;
        this.customConversions20 = 0;
        this.customRevenue1 = 0;
        this.customRevenue2 = 0;
        this.customRevenue3 = 0;
        this.customRevenue4 = 0;
        this.customRevenue5 = 0;
        this.customRevenue6 = 0;
        this.customRevenue7 = 0;
        this.customRevenue8 = 0;
        this.customRevenue9 = 0;
        this.customRevenue10 = 0;
        this.customRevenue11 = 0;
        this.customRevenue12 = 0;
        this.customRevenue13 = 0;
        this.customRevenue14 = 0;
        this.customRevenue15 = 0;
        this.customRevenue16 = 0;
        this.customRevenue17 = 0;
        this.customRevenue18 = 0;
        this.customRevenue19 = 0;
        this.customRevenue20 = 0;
    }
}
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_1'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions1", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_2'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions2", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_3'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions3", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_4'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions4", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_5'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions5", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_6'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions6", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_7'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions7", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_8'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions8", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_9'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions9", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_10'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions10", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_11'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions11", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_12'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions12", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_13'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions13", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_14'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions14", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_15'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions15", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_16'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions16", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_17'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions17", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_18'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions18", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_19'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions19", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM, is_int: true }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_20'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customConversions20", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_21'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue1", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_22'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue2", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_23'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue3", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_24'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue4", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_25'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue5", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_26'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue6", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_27'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue7", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_28'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue8", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_29'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue9", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_30'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue10", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_31'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue11", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_32'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue12", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_33'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue13", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_34'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue14", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_35'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue15", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_36'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue16", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_37'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue17", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_38'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue18", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_39'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue19", void 0);
__decorate([
    (0, model_decorators_1.Column)({ aggregate: model_decorators_1.AggregateFunction.SUM }),
    (0, table_decorators_1.ToTable)(PerformanceTable_1.PerformanceTable, 'field_40'),
    __metadata("design:type", Number)
], VoluumPerformance.prototype, "customRevenue20", void 0);
exports.VoluumPerformance = VoluumPerformance;
