// @paperclicks/optimizer_database/models/entities/Campaign.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleAdsCampaign = exports.AmazonCampaign = exports.TikTokCampaign = exports.FacebookCampaign = exports.Campaign = exports.CampaignCreate = void 0;
const entities_performance_1 = require("../entities_performance");
const traffic_source_performance_1 = require("../traffic_source_performance");
const tracker_performance_1 = require("../tracker_performance");
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const EntityModel_1 = require("../base/EntityModel");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const tables_1 = require("../../tables");
const GoogleAdsPerformance_1 = require("../traffic_source_performance/GoogleAdsPerformance");
class CampaignCreate extends EntityModel_1.EntityModelCreate {
    constructor(model) {
        super(model);
    }
}
exports.CampaignCreate = CampaignCreate;
class Campaign extends EntityModel_1.EntityModel {
    constructor() {
        super(...arguments);
        this.tracker_unique_name = null;
        this.tracker_campaign_id = null;
        this.tr_name = '';
        this.url = '';
        this.managed = true;
        this.duplicate = false;
        this.enabled = false;
        this.deleted = false;
        this.archived = false;
        this.is_whitelist = true;
        this.manual_connected = false;
        this.ga_manual_connected = false;
    }
    get entityType() {
        return 'campaign';
    }
}
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.AutoField)(['traffic_source_account_id', 'traffic_source_campaign_id'], {
        required: true,
    }),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'item_id'),
    __metadata("design:type", String)
], Campaign.prototype, "id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'traffic_source_item_id'),
    __metadata("design:type", String)
], Campaign.prototype, "traffic_source_campaign_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_4'),
    __metadata("design:type", Object)
], Campaign.prototype, "tracker_unique_name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_5'),
    __metadata("design:type", Number)
], Campaign.prototype, "tracker_account_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_6'),
    __metadata("design:type", Object)
], Campaign.prototype, "tracker_campaign_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_7'),
    __metadata("design:type", String)
], Campaign.prototype, "name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_8'),
    __metadata("design:type", String)
], Campaign.prototype, "tr_name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_9'),
    __metadata("design:type", String)
], Campaign.prototype, "url", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_10'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "managed", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_11'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "duplicate", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_12'),
    __metadata("design:type", String)
], Campaign.prototype, "status", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_13'),
    __metadata("design:type", String)
], Campaign.prototype, "traffic_source_status", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_14'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "enabled", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_15'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "deleted", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_16'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "archived", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_17'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "is_whitelist", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_18'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "manual_connected", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_19'),
    __metadata("design:type", Number)
], Campaign.prototype, "ga_account_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_20'),
    __metadata("design:type", Boolean)
], Campaign.prototype, "ga_manual_connected", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_21'),
    __metadata("design:type", Object)
], Campaign.prototype, "ga_view_ids", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_22'),
    __metadata("design:type", Number)
], Campaign.prototype, "bid", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_23'),
    __metadata("design:type", Number)
], Campaign.prototype, "budget", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_24'),
    __metadata("design:type", Number)
], Campaign.prototype, "daily_budget", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_25'),
    __metadata("design:type", Number)
], Campaign.prototype, "live_spent", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_26'),
    __metadata("design:type", Number)
], Campaign.prototype, "payout", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_27'),
    __metadata("design:type", Number)
], Campaign.prototype, "adgroup_cpc", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(entities_performance_1.CampaignPerformance, [['id', 'campaign_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.AYPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "ay_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.ClickflarePerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "clickflare_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.GaPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "ga_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.PublisherPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "publisher_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.VoluumPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Campaign.prototype, "voluum_performance", void 0);
exports.Campaign = Campaign;
let FacebookCampaign = class FacebookCampaign extends Campaign {
};
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.FacebookPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], FacebookCampaign.prototype, "facebook_performance", void 0);
FacebookCampaign = __decorate([
    (0, model_decorators_1.TrafficSourceTypes)('Facebook')
], FacebookCampaign);
exports.FacebookCampaign = FacebookCampaign;
let TikTokCampaign = class TikTokCampaign extends Campaign {
};
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.TikTokPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], TikTokCampaign.prototype, "tiktok_performance", void 0);
TikTokCampaign = __decorate([
    (0, model_decorators_1.TrafficSourceTypes)('TikTok')
], TikTokCampaign);
exports.TikTokCampaign = TikTokCampaign;
let AmazonCampaign = class AmazonCampaign extends Campaign {
};
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.AmazonPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], AmazonCampaign.prototype, "amazon_performance", void 0);
AmazonCampaign = __decorate([
    (0, model_decorators_1.TrafficSourceTypes)('Amazon')
], AmazonCampaign);
exports.AmazonCampaign = AmazonCampaign;
let GoogleAdsCampaign = class GoogleAdsCampaign extends Campaign {
};
__decorate([
    (0, model_decorators_1.OneToMany)(GoogleAdsPerformance_1.GoogleAdsPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], GoogleAdsCampaign.prototype, "googleads_performance", void 0);
GoogleAdsCampaign = __decorate([
    (0, model_decorators_1.TrafficSourceTypes)('GoogleAds')
], GoogleAdsCampaign);
exports.GoogleAdsCampaign = GoogleAdsCampaign;
