// @paperclicks/optimizer_database/models/entities/Content.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Content = exports.ContentCreate = void 0;
const entities_performance_1 = require("../entities_performance");
const traffic_source_performance_1 = require("../traffic_source_performance");
const tracker_performance_1 = require("../tracker_performance");
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const EntityModel_1 = require("../base/EntityModel");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const tables_1 = require("../../tables");
class ContentCreate extends EntityModel_1.EntityModelCreate {
    constructor(model) {
        super(model);
    }
}
exports.ContentCreate = ContentCreate;
class Content extends EntityModel_1.EntityModel {
    constructor() {
        super(...arguments);
        this.enabled = true;
        this.content_type = null;
        this.archived = false;
        this.brand_name = null;
    }
    get entityType() {
        return 'content';
    }
}
__decorate([
    (0, model_decorators_1.AutoField)([
        'traffic_source_account_id',
        'traffic_source_campaign_id',
        'traffic_source_content_id',
    ], { required: true }),
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'item_id'),
    __metadata("design:type", String)
], Content.prototype, "id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'traffic_source_item_id'),
    __metadata("design:type", String)
], Content.prototype, "traffic_source_content_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_1'),
    __metadata("design:type", String)
], Content.prototype, "headline", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_2'),
    __metadata("design:type", Boolean)
], Content.prototype, "enabled", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_3'),
    __metadata("design:type", String)
], Content.prototype, "status", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_4'),
    __metadata("design:type", String)
], Content.prototype, "traffic_source_status", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_5'),
    __metadata("design:type", Object)
], Content.prototype, "content_type", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_6'),
    __metadata("design:type", Boolean)
], Content.prototype, "archived", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_7'),
    __metadata("design:type", Object)
], Content.prototype, "brand_name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_8'),
    __metadata("design:type", String)
], Content.prototype, "url", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_9'),
    __metadata("design:type", String)
], Content.prototype, "image_url", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_10'),
    __metadata("design:type", Number)
], Content.prototype, "cpc", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_11'),
    __metadata("design:type", Object)
], Content.prototype, "additional_fields", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_12'),
    __metadata("design:type", String)
], Content.prototype, "name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_13'),
    __metadata("design:type", String)
], Content.prototype, "quality_ranking", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_14'),
    __metadata("design:type", String)
], Content.prototype, "conversion_ranking", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_15'),
    __metadata("design:type", String)
], Content.prototype, "buing_type", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_16'),
    __metadata("design:type", String)
], Content.prototype, "adset_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_17'),
    __metadata("design:type", Number)
], Content.prototype, "cta", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_18'),
    __metadata("design:type", String)
], Content.prototype, "description", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_19'),
    __metadata("design:type", String)
], Content.prototype, "learning_status", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(entities_performance_1.ContentPerformance, [['id', 'content_id']]),
    __metadata("design:type", Array)
], Content.prototype, "performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.AmazonPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "amazon_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.TikTokPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "tiktok_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.AYPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "ay_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.ClickflarePerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "clickflare_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.GaPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "ga_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.PublisherPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "publisher_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.VoluumPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "voluum_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.FacebookPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], Content.prototype, "facebook_performance", void 0);
exports.Content = Content;
