// @paperclicks/optimizer_database/models/entities/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntitiesModels = exports.Widget = exports.Target = exports.SubWidget = exports.Spot = exports.Site = exports.SectionV2 = exports.Exchange = exports.Domain = exports.Content = exports.GoogleAdsCampaign = exports.AmazonCampaign = exports.TikTokCampaign = exports.FacebookCampaign = exports.Campaign = exports.FacebookAdgroup = exports.AdGroup = void 0;
const Adgroup_1 = require("./Adgroup");
const Campaign_1 = require("./Campaign");
const Content_1 = require("./Content");
const Domain_1 = require("./Domain");
const Exchange_1 = require("./Exchange");
const SectionV2_1 = require("./SectionV2");
const Site_1 = require("./Site");
const Spot_1 = require("./Spot");
const SubWidget_1 = require("./SubWidget");
const Target_1 = require("./Target");
const Widget_1 = require("./Widget");
var Adgroup_2 = require("./Adgroup");
Object.defineProperty(exports, "AdGroup", { enumerable: true, get: function () { return Adgroup_2.AdGroup; } });
var Adgroup_3 = require("./Adgroup");
Object.defineProperty(exports, "FacebookAdgroup", { enumerable: true, get: function () { return Adgroup_3.FacebookAdgroup; } });
var Campaign_2 = require("./Campaign");
Object.defineProperty(exports, "Campaign", { enumerable: true, get: function () { return Campaign_2.Campaign; } });
Object.defineProperty(exports, "FacebookCampaign", { enumerable: true, get: function () { return Campaign_2.FacebookCampaign; } });
Object.defineProperty(exports, "TikTokCampaign", { enumerable: true, get: function () { return Campaign_2.TikTokCampaign; } });
Object.defineProperty(exports, "AmazonCampaign", { enumerable: true, get: function () { return Campaign_2.AmazonCampaign; } });
Object.defineProperty(exports, "GoogleAdsCampaign", { enumerable: true, get: function () { return Campaign_2.GoogleAdsCampaign; } });
var Content_2 = require("./Content");
Object.defineProperty(exports, "Content", { enumerable: true, get: function () { return Content_2.Content; } });
var Domain_2 = require("./Domain");
Object.defineProperty(exports, "Domain", { enumerable: true, get: function () { return Domain_2.Domain; } });
var Exchange_2 = require("./Exchange");
Object.defineProperty(exports, "Exchange", { enumerable: true, get: function () { return Exchange_2.Exchange; } });
var SectionV2_2 = require("./SectionV2");
Object.defineProperty(exports, "SectionV2", { enumerable: true, get: function () { return SectionV2_2.SectionV2; } });
var Site_2 = require("./Site");
Object.defineProperty(exports, "Site", { enumerable: true, get: function () { return Site_2.Site; } });
var Spot_2 = require("./Spot");
Object.defineProperty(exports, "Spot", { enumerable: true, get: function () { return Spot_2.Spot; } });
var SubWidget_2 = require("./SubWidget");
Object.defineProperty(exports, "SubWidget", { enumerable: true, get: function () { return SubWidget_2.SubWidget; } });
var Target_2 = require("./Target");
Object.defineProperty(exports, "Target", { enumerable: true, get: function () { return Target_2.Target; } });
var Widget_2 = require("./Widget");
Object.defineProperty(exports, "Widget", { enumerable: true, get: function () { return Widget_2.Widget; } });
exports.EntitiesModels = {
    AdGroup: Adgroup_1.AdGroup,
    FacebookAdgroup: Adgroup_1.FacebookAdgroup,
    Campaign: Campaign_1.Campaign,
    FacebookCampaign: Campaign_1.FacebookCampaign,
    TikTokCampaign: Campaign_1.TikTokCampaign,
    AmazonCampaign: Campaign_1.AmazonCampaign,
    GoogleAdsCampaign: Campaign_1.GoogleAdsCampaign,
    Content: Content_1.Content,
    Domain: Domain_1.Domain,
    Exchange: Exchange_1.Exchange,
    SectionV2: SectionV2_1.SectionV2,
    Section: SectionV2_1.SectionV2,
    Site: Site_1.Site,
    Spot: Spot_1.Spot,
    SubWidget: SubWidget_1.SubWidget,
    Target: Target_1.Target,
    Widget: Widget_1.Widget,
};
