// @paperclicks/optimizer_database/models/entities/SectionV2.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SectionV2 = exports.SectionV2Create = void 0;
const entities_performance_1 = require("../entities_performance");
const traffic_source_performance_1 = require("../traffic_source_performance");
const tracker_performance_1 = require("../tracker_performance");
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const tables_1 = require("../../tables");
const EntityModel_1 = require("../base/EntityModel");
class SectionV2Create extends EntityModel_1.EntityModelCreate {
    constructor(model) {
        super(model);
    }
}
exports.SectionV2Create = SectionV2Create;
class SectionV2 extends EntityModel_1.EntityModel {
    constructor() {
        super(...arguments);
        this.name = null;
        this.url = '';
        this.cpc = 0;
        this.widget_name = '';
    }
    get entityType() {
        return 'section';
    }
}
__decorate([
    (0, model_decorators_1.AutoField)([
        'traffic_source_account_id',
        'traffic_source_campaign_id',
        'traffic_source_section_id',
    ], { required: true }),
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'item_id'),
    __metadata("design:type", String)
], SectionV2.prototype, "id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'traffic_source_item_id'),
    __metadata("design:type", String)
], SectionV2.prototype, "traffic_source_section_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_1'),
    __metadata("design:type", Object)
], SectionV2.prototype, "name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_2'),
    __metadata("design:type", String)
], SectionV2.prototype, "status", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_3'),
    __metadata("design:type", String)
], SectionV2.prototype, "url", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_4'),
    __metadata("design:type", Object)
], SectionV2.prototype, "cpc", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'field_6'),
    __metadata("design:type", String)
], SectionV2.prototype, "widget_name", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(entities_performance_1.SectionPerformanceV2, [['id', 'section_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.AmazonPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "amazon_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.FacebookPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "facebook_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(traffic_source_performance_1.TikTokPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "tiktok_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.AYPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "ay_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.ClickflarePerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "clickflare_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.GaPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "ga_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.PublisherPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "publisher_performance", void 0);
__decorate([
    (0, model_decorators_1.OneToMany)(tracker_performance_1.VoluumPerformance, [['id', 'item_id']]),
    __metadata("design:type", Array)
], SectionV2.prototype, "voluum_performance", void 0);
exports.SectionV2 = SectionV2;
