// @paperclicks/optimizer_database/models/entities_performance/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntitiesPerformanceModels = exports.WidgetPerformance = exports.TargetPerformance = exports.SpotPerformance = exports.SitePerformance = exports.SectionPerformanceV2 = exports.ExchangePerformance = exports.DomainPerformance = exports.ContentPerformance = exports.CampaignPerformance = exports.AdgroupPerformance = void 0;
const AdgroupPerformance_1 = require("./AdgroupPerformance");
const CampaignPerformance_1 = require("./CampaignPerformance");
const ContentPerformance_1 = require("./ContentPerformance");
const DomainPerformance_1 = require("./DomainPerformance");
const ExchangePerformance_1 = require("./ExchangePerformance");
const SectionPerformanceV2_1 = require("./SectionPerformanceV2");
const SitePerformance_1 = require("./SitePerformance");
const SpotPerformance_1 = require("./SpotPerformance");
const TargetPerformance_1 = require("./TargetPerformance");
const WidgetPerformance_1 = require("./WidgetPerformance");
var AdgroupPerformance_2 = require("./AdgroupPerformance");
Object.defineProperty(exports, "AdgroupPerformance", { enumerable: true, get: function () { return AdgroupPerformance_2.AdgroupPerformance; } });
var CampaignPerformance_2 = require("./CampaignPerformance");
Object.defineProperty(exports, "CampaignPerformance", { enumerable: true, get: function () { return CampaignPerformance_2.CampaignPerformance; } });
var ContentPerformance_2 = require("./ContentPerformance");
Object.defineProperty(exports, "ContentPerformance", { enumerable: true, get: function () { return ContentPerformance_2.ContentPerformance; } });
var DomainPerformance_2 = require("./DomainPerformance");
Object.defineProperty(exports, "DomainPerformance", { enumerable: true, get: function () { return DomainPerformance_2.DomainPerformance; } });
var ExchangePerformance_2 = require("./ExchangePerformance");
Object.defineProperty(exports, "ExchangePerformance", { enumerable: true, get: function () { return ExchangePerformance_2.ExchangePerformance; } });
var SectionPerformanceV2_2 = require("./SectionPerformanceV2");
Object.defineProperty(exports, "SectionPerformanceV2", { enumerable: true, get: function () { return SectionPerformanceV2_2.SectionPerformanceV2; } });
var SitePerformance_2 = require("./SitePerformance");
Object.defineProperty(exports, "SitePerformance", { enumerable: true, get: function () { return SitePerformance_2.SitePerformance; } });
var SpotPerformance_2 = require("./SpotPerformance");
Object.defineProperty(exports, "SpotPerformance", { enumerable: true, get: function () { return SpotPerformance_2.SpotPerformance; } });
var TargetPerformance_2 = require("./TargetPerformance");
Object.defineProperty(exports, "TargetPerformance", { enumerable: true, get: function () { return TargetPerformance_2.TargetPerformance; } });
var WidgetPerformance_2 = require("./WidgetPerformance");
Object.defineProperty(exports, "WidgetPerformance", { enumerable: true, get: function () { return WidgetPerformance_2.WidgetPerformance; } });
exports.EntitiesPerformanceModels = {
    AdGroupPerformance: AdgroupPerformance_1.AdgroupPerformance,
    CampaignPerformance: CampaignPerformance_1.CampaignPerformance,
    ContentPerformance: ContentPerformance_1.ContentPerformance,
    DomainPerformance: DomainPerformance_1.DomainPerformance,
    ExchangePerformance: ExchangePerformance_1.ExchangePerformance,
    SectionPerformanceV2: SectionPerformanceV2_1.SectionPerformanceV2,
    SectionPerformance: SectionPerformanceV2_1.SectionPerformanceV2,
    SitePerformance: SitePerformance_1.SitePerformance,
    SpotPerformance: SpotPerformance_1.SpotPerformance,
    TargetPerformance: TargetPerformance_1.TargetPerformance,
    WidgetPerformance: WidgetPerformance_1.WidgetPerformance,
};
