// @paperclicks/optimizer_database/models/base/EntityModel.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityModel = exports.EntityModelCreate = void 0;
const read_1 = require("../../core/read");
const constants_1 = require("../../core/read/constants");
const model_decorators_1 = require("../../core/schema/model/decorators/model_decorators");
const table_decorators_1 = require("../../core/schema/table/decorators/table_decorators");
const tables_1 = require("../../tables");
const Model_1 = require("./Model");
const GetPerformanceRequest_1 = require("../../core/read/request/GetPerformanceRequest");
class EntityModelCreate {
    constructor(model) {
        Object.assign(this, model);
    }
}
exports.EntityModelCreate = EntityModelCreate;
class EntityModel extends Model_1.Model {
    constructor() {
        super(...arguments);
        this.platform = 'native';
    }
    get type() {
        return this.entityType;
    }
    static async getPerformance(requestArgs) {
        const request = new GetPerformanceRequest_1.GetPerformanceRequest(requestArgs, this);
        const queryBuilder = new read_1.EntityQueryBuilder(request);
        const query = queryBuilder.getQuery();
        query.prettyPrint();
    }
}
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'user_id'),
    __metadata("design:type", Number)
], EntityModel.prototype, "user_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'platform'),
    __metadata("design:type", String)
], EntityModel.prototype, "platform", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.AutoField)(['traffic_source_account_id', 'traffic_source_campaign_id'], {
        required: true,
    }),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'campaign_id'),
    __metadata("design:type", String)
], EntityModel.prototype, "campaign_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'traffic_source_unique_name'),
    __metadata("design:type", String)
], EntityModel.prototype, "traffic_source_unique_name", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, 'traffic_source_account_id'),
    __metadata("design:type", Number)
], EntityModel.prototype, "traffic_source_account_id", void 0);
__decorate([
    (0, model_decorators_1.Column)(),
    (0, model_decorators_1.Required)(),
    (0, table_decorators_1.ToTable)(tables_1.EntityTable, constants_1.ENTITY_TYPE_DISCRIMINATOR),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [])
], EntityModel.prototype, "type", null);
exports.EntityModel = EntityModel;
