// @paperclicks/optimizer_database/core/schema/table/schema/TableSchemaField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableSchemaField = void 0;
const types_1 = require("../../../write/types");
const value_serializer_1 = require("../../../write/value_serializer");
class TableSchemaField {
    constructor(name, table) {
        this.name = name;
        this.table = table;
        this._fieldType = types_1.CHType.String;
    }
    set fieldType(value) {
        this._fieldType = value;
    }
    get fieldType() {
        return this._fieldType;
    }
    setTransformFunction(fn) {
        this._transformFunction = fn;
    }
    transformValue(data) {
        const value = this._transformFunction
            ? this._transformFunction(data)
            : data;
        return (0, value_serializer_1.serializeValue)(this.fieldType, value);
    }
    withSubClassTable(table) {
        const newField = new TableSchemaField(this.name, table);
        newField._fieldType = this._fieldType;
        newField._transformFunction = this._transformFunction;
        return newField;
    }
}
exports.TableSchemaField = TableSchemaField;
